### ISSUE TITLE
Future updates (suggestions/discussion only!)

### ISSUE THREAD

**Issue by BrandonWill** [BrandonWill/GrepolisBot#6]
Hey everyone! As some people may know, I'm not currently playing Grepolis. Therefore, I don't know whether everything is working or not. If something is broken, please create a new issue describing the issue and supply the error if there is one given. That being said, it usually doesn't take more than an hour to fix anything Grepolis will throw at the bot so I have no problem fixing any issues that arises.

This also limits the number of features that I'm willing to implement due to the way Grepolis updates. Automatic researching, no captain farming, etc. are all loaded through the first HTTP Request the game sends with map data, and they change with almost every single update to the game. Because of this, I'm not willing to sit through and read that data to update the bot every time the game updates.

This is mainly a place to address two concerns:

1. Future Grepolis updates that can affect the bot
2. Suggestions for future updates

The current concern that some people have told me about was the new battle point farming. I've looked at it on beta (zz.grepolis) and the data is only slightly different than now. The only thing that has to be changed from the bot's perspective is the mood; there's no more double resources. Outside of that, there isn't anything to be concerned about.

**Comment by beaubeire** [comment]
Maybe add a looting system beside a farming system? Hate when mood of villages drops

**Comment by BrandonWill** [comment]
What do you mean by a looting system?

If you don't want the mood of the villages to drop, you can set the mood to loot at to 100 in the settings.

**Comment by Frank18000** [comment]
a system to have some  standard cities will be great

**Comment by Frank18000** [comment]
an option to program attacks and os

**Comment by nj-costa** [comment]
All Uthar's b0t  features, like... automatic cultural festivals,only add units to the queue when the spells are active, etc. 
Thanks for your work!

**Comment by beaubeire** [comment]
@BrandonWill 

setting mood to 100 doesn't work. You need a system that will demand resources from the villages and not steal resources. You get what i'm saying?

**Comment by BrandonWill** [comment]
@Frank18000 

What do you mean by standard cities? I'll look into automatic attacking next week. It seems pretty straight forward. However, I can't think of a nice way to implement it into the bot. What I mean by that is, how would you like it to be shown in the bot? Where should I put something like that and how should the layout look like (a basic paint picture would suffice if anyone has ideas!)

@nj-costa 

The bot has automatic city festivals right now. The majority of Uthar's features can't be implemented due to the reasons stated in the main post. I'd love to add a lot of features, but they'd break a lot. Given those circumstances and my amount of free time, I'd like to implement other features people would like that don't break almost every update. 

@beaubeire 

This was a program bug from the looks of it. It checked to see whether the mood was below the mood to loot at, which worked for every case unless it was 100. The newest version should have this fixed.

**Comment by Frank18000** [comment]
What I mean is that there is an option to create several templates to their city in terms of troops and buildings. And this template can then load in construction rows

Attack
Maybe an extra button where you get an pop up where you can set te attack time and the cities whom attack and with wat (flyers kolo vs etc). Then select the cities which send os and what. Give a time when the kolo should land. Then a button attack and the bot times the attack en the OS

**Comment by beaubeire** [comment]
@BrandonWill 

Good work, keep it up! :+1:

**Comment by BrandonWill** [comment]
@Frank18000 

Hmm, alright. I'll look into templates.

I'm actually moving away from most of the buttons soon, which is why it isn't really an option. It still has some ways to go (no login button and it has a theme that messes up the alignment), but it'll start looking a lot closer to what Uthar had. http://i.imgur.com/26PsHP6.png

**Comment by Frank18000** [comment]
a tab is also great for atacking. What would be even better is selecting the city ingame that you want to attack and the select the attack modus

**Comment by Frank18000** [comment]
the new bot works great 
good job thx

**Comment by Frank18000** [comment]
to split the farm clock and  the construction clock is also a useful addition I think. It makes detection for grepo more diffecult

**Comment by BrandonWill** [comment]
They're separated, kind of. Farming time is set in the farming tab (if the farming time is less than time set, it won't even open the tab to farm), and everything updates with the time to update thing.

I guess it wouldn't hurt to actually separate them, add a place where you can change their time, and a spot to actually view the timers.

**Comment by Frank18000** [comment]
great
the advantage of splitting is that you can build rows are not filled with singel units

**Comment by beaubeire** [comment]
bot doesn't find farmtowns

**Comment by BrandonWill** [comment]
Do you have captain enabled?

![image](https://cloud.githubusercontent.com/assets/949047/14259793/8b889d8c-fa76-11e5-8f59-e018a06def0e.png)

**Comment by beaubeire** [comment]
Lol, I forgot captain was necessary, never mind!

**Comment by BrandonWill** [comment]
No problem! I have about a month left of classes and then I'll have some time to work on the bot more ^.^

**Comment by beaubeire** [comment]
Suggestion: Choose which towns to farm.

**Comment by BrandonWill** [comment]
Took a bit longer than I thought it would, but it's in the latest version

**Comment by BrandonWill** [comment]
Next on the list will definitely be a town template saver/loader that was suggested earlier

**Comment by beaubeire** [comment]
great work!!

**Comment by beaubeire** [comment]
Quick question though: In the building queu, do i need to add the amount of levels I want a building to go up or do I need to set the level I want a building to be?

thanks

**Comment by beaubeire** [comment]
Also the bot says a city has a conquer, but it actually hasn't

**Comment by BrandonWill** [comment]
You put it the level you want the building to be. 

Conquering still needs to be fixed. It'll say that it has a conquerer if it's sending one out as well. I don't have the time to play a world right now, and then I'd have to wait for one of my cities to start being conquered to write code for it.

I'll probably remove the current code soon

**Comment by Frank18000** [comment]
the  bot is working again thx

only have 200+ city to setup lol

**Comment by BrandonWill** [comment]
If you have a suggestion to make it easier to setup multiple cities, I'll implement it :).

I couldn't imagine going through so many cities to set them all up. Even if I were to add templates, it would take awhile.

**Comment by Frank18000** [comment]
great i can't wait 
one account is 200 + cities

**Comment by icesand** [comment]
Hi the bot works great, Nice job, 
My suggestion is that the bot could store silver to the cave,

**Comment by MrDafisque** [comment]
Hola,

it would be great if you could rearrange the Building-Queue Window - I using the bot on a netbook with a very low resolution and i can hardly see the dropdown menu to pick the cities, because it's almost off screen. Same goes with the ships (last thing i can see is the Trireme)

**Comment by whyalwaysme11** [comment]
@BrandonWill is still required Captain? If yes where can i find it?

**Comment by BrandonWill** [comment]
@MrDafisque 

I'll look into adding a scrollbar for the windows.

@whyalwaysme11 

You can just double click the .jar file now. The .bat file was for previous versions and still allows for more debugging. 

As far as extensions go, they have to be purely written in Javascript for me to import them. I can add features to change to a personal proxy server, but I don't really see anyone using this. Why is a proxy needed to play grepolis?

**Comment by BrandonWill** [comment]
@MrDafisque 

I pushed out 0.19. There should now be a scroll bar for the queue panel. Does it appear/does it help?

**Comment by tepacz** [comment]
Hi!
I started using in on Mac, however i cannot see the icons of the the buildings as well as the units, any idea how to fix it?

**Comment by MrDafisque** [comment]
@BrandonWill Yes, perfect :-) it would also be great if you could add (or unlock) the chrome-feature to zoom out with [Control] + [-] and in with [Control] + [+]. But only if it's not much work, because it's very likly that i am the only one playing on such a low resolution.

Way more interesting would be template for building-queues - i am currently templating by manipulating the buildingsave.txt file :-)

**Comment by whyalwaysme11** [comment]
@BrandonWill Okay now it works, but my problem now is that i don't see the images in the queue so i can't set buildings and the training units. I've updated Java but it doesn't fix it.

The proxy is useful to connect to the same server with different user accounts without using the same IP.

**Comment by BrandonWill** [comment]
@tepacz This issue should now be fixed! Please let me know if this isn't the case. This program was mostly written for Windows and I don't have any way to currently test other OSs.
@whyalwaysme11 This issue should now be fixed! Please let me know if this isn't the case. This program was mostly written for Windows and I don't have any way to currently test other OSs.


@MrDafisque Looking into what I can do about zooming in/out. There isn't an enable option. The browser is rather undeveloped, so I'll have to create my own method to do this. 

Templates are still in development. I probably won't finish these until my summer classes are done in 2 weeks though (they meet everyday so it kills my free time). I prefer to do a lot of testing and implement it correctly rather than doing a rush job that I'll have to rewrite later.  

This may have also been fixed in the latest version. Make sure you're keeping it in the format with the colons like... 00:00:00 because they do matter when saving/loading. However it doesn't matter if it's like 0003:4015:285 as long as the colons are in place

**Comment by BrandonWill** [comment]
Some basic zooming in and out along with the ability to reset the zoom is now in the bot.

    *Zoom in with control and +
    *Zoom out with control and -
    *Reset zoom with control and 0

**Comment by BrandonWill** [comment]
Long overdue for an update! Town templates have been added.

Simply go to the town you wish to save and click save! As always, the tab is always refreshed when you switch back to it. I've only done a little bit of testing, but it seems to work great. When you click load, it should load all the troops, farming, and buildings from the template you selected. I've also fixed some tiny loading bugs that you probably wouldn't have noticed. 

Sorry it took so long. Let me know what you want the next feature to be!

**Comment by Frank18000** [comment]
it works great
the next thing i would like is the atack en def sending option

**Comment by BrandonWill** [comment]
The biggest problem for that will be that I don't play often. Right now to test things, I just start on a new world and I can fix everything. I'll buy captain if farming needs to be fixed. 

In order to get the data to even begin, I'd have to get 2 towns which requires active gameplay or protection from someone on an already established server. I do have google hangouts for the bot; it's githubgrepolisbot@gmail.com if anyone wants to add it to help get attack/defense added. I would prefer a 3 or 4 speed world so I don't have to spend too much time waiting

**Comment by Frank18000** [comment]
i don't understand you completely. Do you need an account to log on and test  with mutiple cities. Or just information?

**Comment by BrandonWill** [comment]
I was planning on doing it on my own account so that you wouldn't be at risk of being banned. It's going to require a lot of testing and work to get it done correctly. I'd like to make it be as human-like as possible to reduce the chance of ban. When I played, alliances had certain times they'd coordinate attacks on other alliances. My idea would be so that you can set up the bot to participate in these without you being there. 

I'd also like to have a toggle to make it not human-like. Maybe you really want to snipe an incoming CS because the city is just that valuable. Outside of creating the GUI, this will be the hardest thing to code because you still don't want the bot to send a bunch of requests to the server. It'll also be the hardest to test because colony ships are so slow

**Comment by Frank18000** [comment]
that sounds great i hope you get it solved

**Comment by BrandonWill** [comment]
Alright, moving forward guys, I plan on only adding new content. By that, I mean if I were to add research, it'd be the new research window below rather than the old one. It'll probably still work when you're like trying to research colony ships, but it might get you banned if you try stuff that isn't in your version of the game.

![image](https://cloud.githubusercontent.com/assets/949047/17310797/a7f01ada-5814-11e6-80eb-8742f098611b.png)

**Comment by beaubeire** [comment]
is the bot working proper again? with battle point villages and stuff.

There is a new world tomorrow on the dutch server and i want a headstart

**Comment by BrandonWill** [comment]
It should be. Keep in mind that if you don't use captain, the risk to be banned is a lot higher. The bot adjusts itself correctly over time, but it isn't perfect and I'm not sure when I'll have the time to do so. 

I've been botting nl46 for almost 2 weeks and it's been running almost all day. Feel free to create a new account and try it out on an older server that does have battlepoint villages though.

**Comment by Daigz** [comment]
Hi BW,

I would like to assist you with the project, but it will take a while until I've gone through the code. If there is any gruntwork you want done in the meantime please just tell me.

**Comment by Daigz** [comment]
First problem already, when i type username and password and select the server, then run the bot a wierd thing happens. The screen starts to blink and I do not see my servers but some others which I do not recognize.

**Comment by BrandonWill** [comment]
Make sure the bot is fully logged in before trying to run it

**Comment by Daigz** [comment]
Hi BW, 

I don't quite understand - I am not even trying to run the bot. I press logon to Grepolis, the dial spins nearly a full 360. Then I am presented with the world KASMENAI (not one of my worlds) and the only one that gets presented and then the page keeps reloading and going between 0 and 10%. Is the bot only working in certain languages/servers?

**Comment by Daigz** [comment]
[console.txt](https://github.com/BrandonWill/GrepolisBot/files/433046/console.txt)

I get quite a few exceptions when starting the batfile (latest JRE). Can this has something to do with it?

**Comment by BrandonWill** [comment]
Hmm. You entered in your username, password, and server?

The server should only be like... nl74
and not the whole address. It works for all grepolis servers

**Comment by BrandonWill** [comment]
Which server are you trying to use?

**Comment by Daigz** [comment]
My mistake, I put in the servername instead of the serverid (as presented on the serverselection page). It works now.. now for some fun with the code, I was hoping for some more comments but one can't have everything ;-)

Are you using Eclipse or Netbeans mayhaps?

**Comment by BrandonWill** [comment]
I'm using IntelliJ IDEA. I'm a student so I can use the pro version for free, but there are other means of obtaining the pro version.

I've been meaning to commentate a lot of the code. Most of the code does need to be rewritten so that it's more clear. There's been times where farming has broken and I'd just throw together some hackfix because I'm too busy to do it the right way.

The current way of farming without captain is honestly the worst thing I could've done.

**Comment by BrandonWill** [comment]
However, the bot is very simple to make. Most of the time is spent in designing the GUI and fixing weird interactions with the game. If you open up chrome and open the developers console (control+shift+I), you just look at the data that's sent and received. If you click on the senate for example, it'll give you all the building data. You can look at addAlerts() to see how this is all handled.

I'll look into commentating most of the code tomorrow. I think the most confusing part in programming features for the bot is to understand that everything is done with concurrency. You have to be very careful with editing data, but this shouldn't be a problem if things are done correctly

**Comment by Daigz** [comment]
Interesting, I do not have much experience with JSON and other webtechnologies, but I started out my career learning JAVA at the university ages ago. When it was all shiny and new. Don't you use a gui builder or some other tool to design the gui?

**Comment by BrandonWill** [comment]
It depends. The browser panel is something I've written. It's using something that has came out recent, JavaFX. For things like the QueuePanel, I use Netbeans to help design them, but I modify a lot of the code afterwards. Pictures have to be rewritten so that they're loaded from the bot's directory and so it'll work on every OS. For instance, I'm working on research right now. Should I have 3 different pictures? Researched, non-researched, and in progress? Or should I just use Java and convert it to grayscale for not researched and yellowish for in progress. I try to put a lot of thought in the design so that it can easily be improved later on. 

If you are interested, you're more than free to add me on hangouts for questions at githubgrepolisbot@gmail.com

The majority of the confusion when working on the bot will be the browser. Oracle (Java) uses a browser called WebView. This is NOT the same browser as Android's Webview. It is missing a huge number of features and things that would make our life so much easier. Oracle's browser has only been out for a year, so I'm still hopeful that it'll improve

**Comment by BrandonWill** [comment]
I plan on converting everything over to JavaFX whenever I have time. It's absolutely beautiful and allows you to even use CSS. You can see it here if you haven't before

http://docs.oracle.com/javase/8/javafx/user-interface-tutorial/apply-css.htm#CHDGHCDG

However this isn't something that'll happen until like next year.

**Comment by Daigz** [comment]
Looks nice, so far I successfully built the bot using IntelliJ IDEA. Now on to get some sense of structure of all the parts in the code.

**Comment by BrandonWill** [comment]
Yeah it's going to be a bit confusing, sorry! Most of the code is inside of Grepolisbot.java. It might help to follow the bot through the bot debugger by looking at the requests as the bot functions. I start work in an hour, but I'll try to respond when I can if you're curious about anything specific.

**Comment by Daigz** [comment]
Suggestions after looking over it a bit..

**Setting operational hours**

It should be enough with two active spans a day.

Mon Activehours 1 [00:00] to [00:00] Active Hours 2 [00:00] to [00:00]
Tue Activehours 1 [00:00] to [00:00] Active Hours 2 [00:00] to [00:00]
Wed Activehours 1 [00:00] to [00:00] Active Hours 2 [00:00] to [00:00]
Thu Activehours 1 [00:00] to [00:00] Active Hours 2 [00:00] to [00:00]
Fri Activehours 1 [00:00] to [00:00] Active Hours 2 [00:00] to [00:00]
Sat Activehours 1 [00:00] to [00:00] Active Hours 2 [00:00] to [00:00]
Sun Activehours 1 [00:00] to [00:00] Active Hours 2 [00:00] to [00:00]

Note: If the shift is entered as 00:00 to 00:00 the shift is disabled.

**Email notifications from the software (capchta etc, attacks etc.).**

** Possibily to farm/raid inactives just as villages.**

**Comment by Daigz** [comment]
Nah, not a problem BW. I'll take my time trying to get my grip around this :-)

**Comment by BrandonWill** [comment]
Ah yes, but then you have to get into time zones :). Do they want it for their time at those hours or the time they see in game? When we make requests, we only get told the current time in milliseconds for UTC, which is usually neither of those. 

Attack/Defense has been suggested before, but they'd have to be given a higher thread priority. If you have 200 towns, you'd want the bot to stop farming and building. It'd switch to the town, send off the attack(s)/defense(s). I haven't looked too much into those though; I'm mostly concerned about how I'd be able to determine travel time.

**Comment by Daigz** [comment]
I have no intention of pushing you to become a full time bot developer so I am more thinking using the KISS principle.

Working hours is local client time to emulate human behaviour i.e. sleeping at night and such.

Regarding the farm/raid inactives I didnt mean a battle-bot, just that sometimes you keep on attacking inactive for resources, once you emptied them it tedious to send a raiding party every 8 or 12 hours. Just a simple interface where you specify originating town and target and ~timespan and units to send. Much like farming.

**Comment by BrandonWill** [comment]
True. I guess this hasn't been too big of a concern of mine. I've ran the bot for more than 6 months previously 24/7 before my alliance won the server. If a moderator looks into your account, they can see all the requests that we make and you'd be banned regardless. Moderators are honestly their only anti-cheating device. I've sent a large number of requests that all received server errors without getting banned. Heck, for the past 3 months until last week, the bot never even switched towns before building in them (this has been fixed lol).

True, but sometimes towns can ghost and inactives can disappear. A new player could have the same townID, and the bot could try to send some weird 48 hour attack. This would require reading map data to ensure that neither of those are happening.

**Comment by Daigz** [comment]
Regarding the distance I have written a lot of tools for other games, and the difference between Grepolis and most others is that this seems to be a bit easier since its pixelbased instead of hexagonal maps.

https://en.forum.grepolis.com/index.php?threads/how-to-convert-distance-into-travel-time.38691/

**Comment by Daigz** [comment]
"True, but sometimes towns can ghost and inactives can disappear. A new player could have the same townID, and the bot could try to send some weird 48 hour attack. This would require reading map data to ensure that neither of those are happening. "

As you can tell I am a noob at this game, I am basing my suggestions from other games I have written tools for, where they have a lot of other automated stuff to detect bots.

**Comment by Daigz** [comment]
I am mainly interested in this for the .. how should we put it.. the challange since creating tools is a nice harmless hobby :-)

**Comment by Daigz** [comment]
And for the farming stuff, restrict it to the same island as the originating town. You have to manually put in the town to attack. Name/id or such.

**Comment by beaubeire** [comment]
Nice, a little bit of assistance!

**Comment by BrandonWill** [comment]
@Daigz You're more than free to implement any feature you'd like. Most of my updates for the next year will be to fix the bot if it breaks. I'm working full time and a student full time, so my free time is extremely limited and Grepolis isn't a top priority.

**Comment by BrandonWill** [comment]
I've pushed out 2 updates this month in case people missed them. They're nothing too big; just annoyances I've noticed while letting the bot run this month.

* Fixed a bug that caused the current city to not be the same if you left the queue panel
* The bot will now rush lumber camp 15 and silver mine 10 so that you can build docks faster. Before, you'd have to wait until the town was almost fully built.
* Fixed the bug where you'd have to restart after loading a template
* Backend changes. Working on changing over to JavaFX!

**Comment by Frank18000** [comment]
it is working great
i hope that the attacking option is coming soon

**Comment by BrandonWill** [comment]
My desktop's harddrive crashed and I lost the files to the GUI creator, so I can't really add anything to the Queue/farming tab without rewriting them. However, I was only going to add some small things like how a lightship/bireme city doesn't need to open up the barracks at all. 

With that being said, the attack option is very appealing now that I'm playing again. And I don't have to worry about finding how to calculate distances on my own (grepolis gives it to you). It'll probably be the next feature that's added when I have time

**Comment by Frank18000** [comment]
i can't wait for the moment you have the attack  option working. 
sorry that you lost a drive. I have a nas in my home with several drives in it and a back up system zo no los of data

**Comment by BrandonWill** [comment]
I keep most of my stuff on the cloud, so it wasn't too big of a loss. Most of the time is going to go into designing the user interface to this thing. I'm still trying to decide on how it'll look because I'll want to see the timed attacks and the troops that are being sent with them, and I'll want it to be editable (I have a bad habit of entering in the wrong numbers).

Anyways, I pushed out a new update. Due to my shady internet recently, the bot will now restart if it encounters an error

**Comment by Frank18000** [comment]
A cloud is the other option to backup stuff.
What kind of UI are you thinking of? The things you want to see and correct before you  push the send button are wright i think. The bigest problem i think of is the anti timing

**Comment by BrandonWill** [comment]
At first, I'm going to try to make it as simple as possible lol. Having a basic attack tab is the current plan. Here's the current mockup in paint; nothing has been programmed yet.

![image](https://cloud.githubusercontent.com/assets/949047/19045565/81bee98e-8967-11e6-901c-ddbc545aec94.png)

There will also be a spot to load up the travel times of your troops.

I may be able to scrap the time converter tool (and I really do want to). I think I remember seeing the chosen timezone given somewhere by Grepolis. It would make it a bit easier to create

**Comment by Frank18000** [comment]
it looks right 
i think that the antie timing could be anoying too

**Comment by BrandonWill** [comment]
Yeah it can be. I plan on just doing -11 for all the attacks before a colony ship and hoping for the best. Plus this will also allow you to sneak in a colony ship in the middle so it can't be sniped a lot easier.

**Comment by Frank18000** [comment]
but then there is a risk that def get there first and it will kill the kolo.
Is it possible to make the bot check if the atack and def is alright and when it is not okay the bot will cancel the def

**Comment by BrandonWill** [comment]
The anti-timer is + and - 10 seconds. 

I guess the bot could look into it. This is just going to be an attack tab for now; not a defense one. I could see it being a lot more important for defense

**Comment by Frank18000** [comment]
you are right but sending defense after an kolo is just as imporant lol

**Comment by BrandonWill** [comment]
Ah, that's going to have to wait for the defense tab! It isn't as important on revolt worlds.

**Comment by Frank18000** [comment]
on a revolt world it is not important to send os but on a conquer world it is. Is it possible to make it a choice: send as def and send as attack?

**Comment by BrandonWill** [comment]
I'll try to get them both added eventually. More than likely it'll require me to be on my Christmas break before I even get started.

After almost 4 months of suicide botting (running 24/7) and joining a server 2 months late, my account was finally banned. I managed to become one of the top 10 players in my ocean along with being in the top 200 fighters for the world. Since it took so long, I assume it wasn't from detection but rather a moderator looking into my account. I don't see a date, so I assume it's permanent. Not too bad of a run 

![image](https://cloud.githubusercontent.com/assets/949047/19620074/b17dcca0-9842-11e6-828c-ee9b3b607c97.png)

**Comment by Frank18000** [comment]
lol i can´t wait for chrismas

**Comment by jokerO0o** [comment]
is it not working anymore?

**Comment by Frank18000** [comment]
on the worlds i play it i working

**Comment by jokerO0o** [comment]
how do you use it? i mean i put my details. word name, and the press log in i cannot see how to use  build queue or farm :/

**Comment by Frank18000** [comment]
in the tab queue u can set for every city troops and buildings. push save under the cityname and after every thing is set push the starbutton in the browser tab. In the log tab you can see what happens.

**Comment by jokerO0o** [comment]
https://gyazo.com/cb00dcbeff650dc8b73c3724f5f032ab

https://gyazo.com/a2c7e04eeb448fabcb60d301a78bda69 

it doesnt even recognize my cities :(

**Comment by BrandonWill** [comment]
What server are you trying to play on? It seems to be working fine for me.

![image](https://cloud.githubusercontent.com/assets/949047/20844801/99af15d0-b88f-11e6-9bcc-753e3e78e7c4.png)

**Comment by Frank18000** [comment]
every body a merry Christmas

I hope that you get the attack option running

**Comment by BrandonWill** [comment]
Sorry for the late response, I was working on the attack option during December but it's nowhere near completion. 

Due to my heavy class-load and work, I don't think I'll be able to finish it until after I graduate in May. Since I'm not currently playing the game, you guys will have to let me know if something stops working for me to fix it.

**Comment by Frank18000** [comment]
school is more important

**Comment by beaubeire** [comment]
Does the farming feature still work?

**Comment by Frank18000** [comment]
yes. I yust started on a new world and it works

**Comment by Darkosz** [comment]
Hello guys, everything works properly for me but i have one question. Sometimes the bot says "that he farming village "Gak\u00fc" isn't ready to be farmed or is farmed to its' cap!" Which is okay the problem is that it's not true. The it hasn't been farmed to it's cap and it's ready to be farmed. Still it doesn't farm all of them the funny thing is that it's farming 1-2 from the 4.

**Comment by Darkosz** [comment]
Like this, although the first 2 has lower cap then the last 2 it's still farming these ones.

16:37:56 "Roszpszirosz" has been successfully farmed!
16:37:57 "Roszgi" has been successfully farmed!
16:37:58 The farming village "Aegavgavna" isn't ready to be farmed or is farmed to its' cap!
16:37:58 The farming village "Hipszi" isn't ready to be farmed or is farmed to its' cap!

**Comment by Darkosz** [comment]
It's looks like the bot stops farming it if it's doesn't higher then lvl 2. But im not sure.

**Comment by BrandonWill** [comment]
Hmm that's strange. I might've messed up a calculation somewhere.

I've pushed out an update that'll give a more detailed report on farming villages and why it may not be farming them

**Comment by Darkosz** [comment]
Okay, i will update it.

Other question and maybe suggestion also. Can you make the bot prioritize the buildings over the units? Like it's pretty hard to manage it like how it works now. I can only allow the bot to build buildings so it's doesn't spend all of the resources for units, which i have to build manually.
Like the bot only spends what he doesn't need for the next building it's plan to build.
Or give an option where we can deicide which one do we want to prioritize.

**Comment by BrandonWill** [comment]
The bot constructs buildings before attempting to build units. This approach works really well on higher, 3+, speed worlds because it'll never run out of resources from farming and it has no problem keeping both queues full. However, I did play some lower speed worlds and it does really suck in that department. 

The problem with your suggestion is that the bot doesn't know the amount of resources it'll cost to build a building. The game sends back a response for every building actually telling us if we have the resources to build it. This is the same for troops. Because of this, the bot doesn't actively track the number of resources the town currently has. 

I've attempted to grab this data for informational purposes before, but most of the methods result in a ban or give outdated numbers. The last method I tried actually ended up creating fake towns (grepolis was actually giving us resources for towns we didn't own. The bot added them and then tried to read it's town data and everything. Whoops). 

The older grepolis bot had a separate timer for buildings and troops. Would adding a separate timer so you can set how often it builds troops help?

**Comment by Darkosz** [comment]
Ohh i see, that's lame a bit :/

Yea that would be awesome!
Like only build units after 3 hours or something like that.
It would solve the problem. thanks for the fast response Brandon u are great! :)

**Comment by Darkosz** [comment]
Other thing, should i do anything special for GRC to work? Cause simply with the button on the script it doesn't work for me.

**Comment by BrandonWill** [comment]
Alright. I'll see if I can throw it into the current GUI so you can have it sooner.

I just checked and it reads the script correctly from the URL. It's possible they added something new that isn't supported by the java browser. This was the only script that only use Javascript when I checked; the other scripts used some of tampermonkey's API. This also may have changed

**Comment by Darkosz** [comment]
Not that big of an issue, the only thing is miss is the alarm from it.

**Comment by beaubeire** [comment]
A nice feature you maybe can add is that the bot automatically completes a building for free @ 5 minute mark, is that possible?

**Comment by Croven1** [comment]
Dear BrandomWill,

I'm also a developer and I'm interested in collaborating into making a better app. Is there an handy way to communicate with you please ?

Thanks in advance!

**Comment by BrandonWill** [comment]
@beaubeire What do you mean? The bot isn't able to cheat and automatically complete buildings for free if that's what you're asking. 

@Croven1 
You're more than free to email me at githubgrepolisbot@gmail.com

**Comment by beaubeire** [comment]
@BrandonWill 

I'm meaning that every "instant buy" feature becomes free if the construction time goes below 5 minutes. This will work for upgrading/demolishing buildings and researching in the academy, but not for recruiting units and healing heroes. So maybe you can add a feature so the bot automatically will "instant buy" a building or research when the construction time goes below 5 minutes.

**Comment by BrandonWill** [comment]
Hey everyone, I've pushed out the first bot update in awhile. I've included the GSON library in the bot and it's the first time I've ever distributed an outside library, so let me know if you run into any errors! The library will make it easier to parse through a lot of the data that confuses the bot. Right now, only farming uses it. I didn't want to convert everything until I knew it worked for everyone. 

With the GSON library, captain-less farming should now be nearly flawless. The bot doesn't try every farming village for all your towns anymore. I also updated the values for farms. Grepolis buffed them and allowed people to farm more resources and at a faster rate, and the bot now reflects that. 

The bot now opens your academy on the first run through of your towns. For now, this is only to help captain-less farming because it doesn't know if some things are researched. If you're doing captain-less farming and you research villager's loyalty, then you have to restart the bot afterwards for it to notice and farm correctly.

@beaubeire I'm looking into this now. However, it uses the same method that finishing off the building with gold does.

**Comment by Frank18000** [comment]
i have instald it and it works perfect. 
Is there any progress on yhe atttaking option?

**Comment by BrandonWill** [comment]
I've updated it again. The bot will now automatically finish buildings if they're below 5 minutes. However, there's a catch... the bot must've put it in the queue in the first place.

@Frank18000 Yup! It's still in the early stages. I ran into some issues with having the correct towns, but the GSON library will help with debugging. With how little I'm paying attention to the game, I'm probably going to end up testing it on the bandit camp from the tutorial. This also means there's a possibility that there will be a checkbox you can click that'll automatically do all 100 camps for ya.

**Comment by Frank18000** [comment]
okay sounds great
i hope to see it soon this update

**Comment by Frank18000** [comment]
i have instaled the new version know i have to wait if the 5 minute features works

**Comment by BrandonWill** [comment]
Right now, it doesn't really work if you use the adviser since it doesn't take into account that the previous buildings were moved ahead 5 minutes. I'll try to look into a fix for this

**Comment by beaubeire** [comment]
@BrandonWill 

Good job, keep up the good work!

**Comment by Darkosz** [comment]
Can you make a cooldown for the bot on relogging? Cause it's annoying that i can not ask for help from my phone cause the bot keeps relogging.

**Comment by Darkosz** [comment]
Like, when i login from my phone the bot gets an error and stops working.

**Comment by Darkosz** [comment]
Okay so i would like 2 things.

One of them is the error thing. So if i use my account from my phone and the bot running on my computer then the bot stops with an error. Can u fix it somehow?
The other thing is the festivals. I would like to optimize them, like sometimes i want to save up resources but the bot keeps spending it on festivals.

**Comment by BrandonWill** [comment]
@Darkosz the bot should only error out if it's logged out while it's trying to do something such as farm. I can stop it from reconnecting if the bot is idling, but most of the errors are because the bot can continue to run even if grepolis updates. It's a safety net to ensure that you're not banned if they change anything. Sadly the last time I tried to get the bot to restart automatically, it ended up with a ban. However, the bot can now load farming villages a different way so I may be able to get it done correctly

How would you suggest saving resources? The only way the bot can know if the town has resources is by checking the farming villages. Not everyone uses the farming features, and the towns on the islands without farming villages wouldn't be compatible. Much like finding data for the building queue, I haven't found a good way to obtain accurate data without greatly increasing the chance of being banned. The data is available, but these are not calls that a normal player would ever make. Whenever I have free time, I plan on rewriting what I can into the new library I'm using. I'll check to see if it's possible.

**Comment by Darkosz** [comment]
I think u misunderstood me Brandon.

I don't want the bot to save up resources. I want to be able to have an option to turn off the automatic festivals. Because sometimes i need the resources that the bot spends on them. But Still, keep up the good work! :)

**Comment by Adriskye** [comment]
00:57:50 (city name) has a conqueror! Skipping town! 
why that error? i dont have it :/

**Comment by BrandonWill** [comment]
@darkosz I'll look into adding a button for this. I was going to add a culture tab originally, but the only thing we can automate is the city festivals sadly. They added protection to everything else when they rewrote the agora awhile back. 

@Adriskye Sorry, I don't see when you edit the comments over posting again. The conqueror code is old, but I reverted back to an old copy because a feature I added caused myself to get banned. I'll push an update out to remove this when I get off of work in a out 8 hours. 

Do you know what the bot was attempting to do when you received that error? That's related to the new farming library I added. Is farming currently working for you? There may be a chance you simply received an error from the server and the bot tried to read the farming data from it

**Comment by Adriskye** [comment]
It was just trying to update the academy , didnt used other that moment .

btw will u add it as a program or something? it use a lot of ram with java :/ it would be amazing

**Comment by BrandonWill** [comment]
The bot probably won't ever leave Java. The browser is needed currently in the code because of the way it's written. However, I believe I can rewrite the way it handles responses so that the bot only needs the browser to log in and then can be disposed of. This way the bot won't have all of the memory usage. The bot only needs the data the server sends at the beginning to load the towns and farming villages. 

However, it does make the bot more vulnerable to detection, but I don't think that's an issue I have to worry about right now.

**Comment by Adriskye** [comment]
Maybe a scrip for tampermonkey like autofarmbot would be a good way , i rly dont know but with java app it will deal a lot of errors , doubt it will be detected easy . btw it should add something to resolve captchas to dont be vulnerable like u mean. eventho gj dude with ur yob

**Comment by BrandonWill** [comment]
It would take too long to rewrite the whole bot in another language. The bot detects captchas if there are any. However, I haven't seen one in over 2 years on any of the newer servers, so I haven't had any reason to spend time adding a captcha solving service.

Their methods of detection take awhile to kick in, which is why it takes awhile for me to release some updates. It took me over a week to get banned calling a method I shouldn't have before.

**Comment by Adriskye** [comment]
Since last update (today) it dont find villages to farm :/

**Comment by mig9284** [comment]
impossible add GRC

**Comment by BrandonWill** [comment]
@Adriskye I started up the bot and I don't have this issue. Is it still one?

@Darkosz 
@mig9284 
This feature has been fixed. Everyone will now have a new "localStorage" folder. The browser used to generate this automatically. It stopped doing it in a later Java version I assume

**Comment by Adriskye** [comment]
It´s not fixed for me , i just downloaded the new patch but nothing :/
13:09:13 Towns with farms found: 0

**Comment by BrandonWill** [comment]
Which server are you playing on? And are you using captain?

**Comment by Adriskye** [comment]
I dont use captain , en93

**Comment by BrandonWill** [comment]
Are you sure you don't have captain enabled? It will only show that message if it detects it as enabled. Also, is the time on your computer correct?

Here's how it should look starting up. I tested this on en96.

![image](https://cloud.githubusercontent.com/assets/949047/24792294/6669e276-1b4b-11e7-9654-c4b77a297d01.png)

**Comment by mrThoZ** [comment]
![bot](https://cloud.githubusercontent.com/assets/15823778/24798877/b8549ac0-1b98-11e7-9382-d2feca858e2b.JPG)
Same issue, Captain is enabled.

**Comment by BrandonWill** [comment]
Are you waiting for the bot to completely login before starting it?

**Comment by Momanus** [comment]
Hello, Can a Raspberry PI handle the bot ?

**Comment by BrandonWill** [comment]
I don't think Raspberry PI can handle the bot. JavaFX's webview client is pretty intensive for Grepolis. 

If they released a Raspberry PI that has 2 GB of RAM, it'd be much more likely to run the bot

**Comment by lusuma** [comment]
Good program and excellent work. Congratulations.

I think it is very necessary/urgent what  Darkosz on 13 Mar said "I do not want the bot to save up resources." I want to be able to have an option to turn off the automatic festivals. But still, keep up the good work! :) ". Resources are spent unnecessarily at festivals when they may be needed for other things.

**Comment by lusuma** [comment]
Good job. Thank you very much for your effort.

**Comment by BrandonWill** [comment]
Version 0.22 is now out. I have added the option to turn city festivals on/off in the main settings tab. 

Since it's the only one that can be easily automated still, I doubt I'll be adding a culture tab anytime soon.

**Comment by BrandonWill** [comment]
Hey everyone, I'm finally getting around to doing the UI rewrite so I can make changes a lot easier down the line. Since I'm writing it all from scratch, it's something that will take awhile. Here's what I've done so far and the comparison between the two.

![image](https://cloud.githubusercontent.com/assets/949047/26235170/68446a96-3c37-11e7-9b06-8c042f6ebe8f.png)

I went with textfields over spinners because the spinners just take up too much space.

I'm thinking about changing up the way buildings look as well. I copied the design from the previous Grepolis Bot, but I think the buildings would look much better if they had the same layout as the troops do (a 40x40 picture with the level below it rather than in the top left corner). I'll probably keep the spinner for buildings

**Comment by Frank18000** [comment]
it looks great
the attacking option is that also in the making

**Comment by BrandonWill** [comment]
Yup! The old troop section before was just part of the queue panel and it couldn't really be used elsewhere easily. I rewrote it all so that the whole troop section is a panel itself (it has a black border to show you it's area). This allows the code to be easily implemented elsewhere, such as the attack panel. 

Whenever I have another free day, I'll probably be able to implement a very simple one

**Comment by Frank18000** [comment]
great i can not wait to see it

**Comment by BrandonWill** [comment]
I pushed out an update with some things still being developed in it; it's mostly backend changes.

If you refresh the game page by right clicking after you bought captain, the bot now recognizes this and will farm with captain.
If captain is going to expire within a minute, the bot will now switch to farming regularly. This doesn't require a page refresh.
The bot now detects your current timezone settings in game when you log in. I've noticed that the game's time may be off by a couple of seconds. I'd like some feedback to know if this is the case for everyone or just me before moving forward with working on attacking

**Comment by Frank18000** [comment]
18:12:57 Verify server time is: 18:12:57 to me it looks alright

**Comment by lusuma** [comment]
Important server time error:
Verify server time is: 18:34:21
Real time server: 18:32:51. Server es61

**Comment by BrandonWill** [comment]
Lusuma, can you make sure your click is in synch? In Windows, you can right click on the clock to adjust the time/settings via the internet. It should look like this:

![image](https://cloud.githubusercontent.com/assets/949047/26415521/c295731c-4080-11e7-85d9-1bc0b7126393.png)

You would just click "update now"

**Comment by lusuma** [comment]
It was already configured as you indicated.
However after performing a manual synchronization, it works correctly.

It happens when you have not synchronized for days
![reloj](https://cloud.githubusercontent.com/assets/28656917/26417591/ea00f17c-40b9-11e7-95ae-ac44a93da66d.jpg)

**Comment by Frank18000** [comment]
19:54:49 Verify server time is: 19:54:49

**Comment by BrandonWill** [comment]
@lusuma I was mostly curious to see how close the bot itself can generate the time to the server standings. The time on your computer is always bound to be off by a bit. You can probably see why creating an attack method will be a pain to create. 

The game does give the current server time in unix time upon login. It wouldn't be too hard to calculate the offset and always account for it. But since computers keep track of time through vibrations... this may need to be constantly updated.

@Frank18000 The bot always seems to run perfectly on your machine ^.^

I'll continue to work on the attack module, but it does look like there will be a lot of bugs I'll have to fix for different people after I first release it.

**Comment by Frank18000** [comment]
why not use the grepo time and not use the computer time. Instead of trying to compensate for the difference?

**Comment by Frank18000** [comment]
or do an refresh 5 minutes before the attacks are send and use that time to time on

**Comment by BrandonWill** [comment]
Sadly, a time conversion will still be necessary so it isn't confusing to you guys. Their server time is in UTC.

The problem with that is the bot will have to keep track of time. It'd be pretty resource intensive to track time at the millisecond level. I'll keep doing testing and refreshing for awhile. I'm curious to see if it will lose seconds while the bot is running

**Comment by lusuma** [comment]
Brandon, what program do you use to modify the source code? I have some ideas and if they work, give you the modified code lines.

**Comment by Frank18000** [comment]
Brandon. Is it possible to displa the number of free villagers on the build tab

**Comment by mmindes** [comment]
When I log in grepolis tells me invalid world, how do I fix this?

**Comment by BrandonWill** [comment]
@lusuma I use Intellij, but netbeans also works.

@Frank18000 Yes, this information is available. Where should I put it at in the bot.

@mmindes When you log into grepolis in your browser, it will say the server name (such as en) and the world number (such as 98), you'd put this in settings tab as en98.

**Comment by mmindes** [comment]
In the beta only works on worlds en?

**Comment by BrandonWill** [comment]
It works for all worlds, you just have to change it from en. You can change it to us or whatever you want.

**Comment by Frank18000** [comment]
![image](https://cloud.githubusercontent.com/assets/16511894/26519804/14170f78-42c7-11e7-80cf-28dbbbbd175f.png)
somthing like this

**Comment by BrandonWill** [comment]
It probably be useful to list the city name since it's listed on the right. The total number of citizens probably wouldn't be too useful since buildings take up a lot of population and the city name is on the right as well.

![image](https://cloud.githubusercontent.com/assets/949047/26524356/5005cb32-42fe-11e7-8476-e7d3e03e1455.png)

Whenever I get around to finishing up the queue tab rewrite, I'll add it to the bottom left where you have it alongside the checkboxes to enable/disable barracks/docks/senate building to speed up the bot.

**Comment by Frank18000** [comment]
great thx

**Comment by radekast** [comment]
Hello, Czech servers can not log in - server cz37

**Comment by BrandonWill** [comment]
@radekast I think this is a problem with Windows 8; I'm not sure why. After it enters in your username/password, right click on the page and just click refresh. It'll log you in correctly.

**Comment by radekast** [comment]
Super, thank you, I'm using windows 10 Enterprise

**Comment by isxs** [comment]
Do you recommend using proxy/VPN, or another hiding method - i mean to hide real ip if i will be banned.

**Comment by atscoville** [comment]
Can you add the DIO Tools Same as the GRC script? The DIO Tools has an option for Scroll built in and many useful things that would be a great addition to the software.

**Comment by BrandonWill** [comment]
@radekast Hmm. I'm not sure why different Windows operating systems are acting differently. It seems to be a recent problem though because I have used my laptop in the past without it being an issue. However, it is one now.

@isxs If you wish, you can hide your IP. I believe I've been the only person that's been banned while using the bot though. Nobody has reported being banned the 2+ years it's been out :). And it was because I was testing some features I wanted to add. Although I recently stopped playing a server, I ran the bot 24/7 for nearly 3 months. I can assure you that they don't ban IPs nor is it flagged to determine if you're still botting. The account that I was botting on for 3 months was banned about 3 years ago for botting (a mod found my account and it was the c# version of this bot mentioned in the readme) and again a year ago from my personal testing. If a mod does manage to look into your account and you get banned, then I do suggest that you do stop botting if you want to continue the world.

**Comment by Frank18000** [comment]
Dear Brandon,

Is there some development or are you to busy in real live?

Frank

**Comment by BetaWing** [comment]
The Bot randomly crashes from hour to hour idk why. Hopefully brandon shows up to solve the crash thingy :(

**Comment by matz98** [comment]
hey, the bot is only crashing at me, when i change someting in the browser, so may pause him before klick anything. and i would like to see an automatic trader, for example, between own cities or in the marketplace. if one town have to much wood and another not enouth i send away the recoures to the other city. another idea would be an auto silver to cave manager. if you have to much silver in an city and all the other cities also have enout of it, so i get into the cave.

**Comment by matz98** [comment]
one more, is it possible to update the browser interface more often? because i dont see the real amount of recources or troops.

**Comment by BrandonWill** [comment]
@Frank18000 There hasn't been any recent developments on the bot. 

@BetaWing It's more than likely your internet connection killing it. The bot just dies if that's the case. I've been able to run it for awhile now.

@matz98 
1. I can make it so it won't automatically select the world if the bot is paused. However, it will require you to select the world yourself even if you unpause it. The bot mostly acts as information is provided by the game; it rarely forces the game to tell it anything. As for the other suggestions, it's very unlikely that I'll add a big update such as that at this point in time. Development has stalled more into keeping it running rather than adding new features for now.

2. It's not possible to update that information. Most of the requests the bot does happens behind the scenes. This means you don't see the correct information on the browser, but the bot sees everything.

**Comment by matz98** [comment]
all right thank you

**Comment by Frank18000** [comment]
I hope yoe have the time and the will to do some further development on this project.

everybody good whishes for the new year

**Comment by BrandonWill** [comment]
I've pushed out another update to version 0.22. This will be the last update that will be compatible with Java 8. As Java 9 is now released, future bot releases will require it due to the many changes that they've added.

**Comment by BrandonWill** [comment]
0.23 has been released for Java 9. I will keep the previous version around for awhile but it won't receive any future updates.

**Comment by NeveuL** [comment]
Hello. 
1. Do you still update this bot?
2. Is it undetectable?
3. Why don't you do web script?
4. Is it keylogger?

**Comment by BrandonWill** [comment]
@NeveuL 

1. In some sense yes. I try to keep it running.
2. I hope so? Nobody has mentioned anything about bans. In the past, I've ran it for more than a year nonstop. Only bans I've heard of are from myself testing new features.
3. I converted an older bot from C# to Java. I have no intention of changing it over to JS.
4. The bot is open source. Feel free to look through it and compile the source yourself.

**Comment by NeveuL** [comment]
Thanks for your answers. You are the best.

**Comment by BrandonWill** [comment]
Automatic researching is currently in progress. I have finally gotten around to some of the user interface redesign I mentioned before. Here's a little bit of a preview. You will be able to click on the picture to tell the bot what to research and the image won't be gray anymore. It's actual in game status will be below it.

![image](https://user-images.githubusercontent.com/949047/35186421-d483b704-fde1-11e7-902a-80807c5729b3.png)

**Comment by Frank18000** [comment]
it looks great.

**Comment by BrandonWill** [comment]
I was hoping to finish this feature this weekend, but it's going to have to be pushed back until I have more than 1 town to confirm everything.

Current features:
* Bot recognizes all of the academy researches and loads them
* Bot displays if they're actually researched in the UI
* Clicking the UI does change the picture and change the status in the backend that it should be researched. Currently this only works for Slinger, but having 1 working is the hard part here.
* Saving and loading works between opening/closing the bot
* Currently, clicking on Slinger will set it to be researched. Clicking "save" for the town isn't necessary for the academy. Possibly change troops/docks/buildings to act the same way and remove the save button in the future.

Missing features:
* It doesn't actually research anything yet (code is there)
* Code is missing for checking the academy queue before researching. This is needed before the above step
* Everything but Slinger still needs to be written for UI. I started on the whole row, but they're missing a lot of functionality still.

----
Picture below to give an update and show a bit of what I mentioned in features.

![image](https://user-images.githubusercontent.com/949047/35197498-f26b9d64-fead-11e7-8979-1d8d7782bce6.png)

**Comment by BrandonWill** [comment]
The UI is now done.

Missing features:

*It doesn't actually research anything yet (code is there)
*Code is missing for checking the academy queue before researching. This is needed before the above step. I will have to create this method and have the bot track all researching. This will take some time as well. The game tells us if barracks/docks/buildings queue is full or not. It doesn't tell us for academy queue so I have to write this portion from scratch.

![image](https://user-images.githubusercontent.com/949047/35545356-c18c2cb4-053c-11e8-8348-5f6a7e406a8f.png)

**Comment by BrandonWill** [comment]
Testing in progress

![image](https://user-images.githubusercontent.com/949047/35604952-5452554c-0613-11e8-9e75-4549f66ae66e.png)

**Comment by Dracks977** [comment]
thx for work guys

**Comment by NS58** [comment]
Hey Brandon, first of all, great work on this bot, I love it, and have been using to problem free to just farm for the past 2 months.
Just a few things I would like to see in the bot when you have time to add them, if you are still working on this bot:
To save time in the loop, would it be possible to not check the troop and building queues? Since I never use these 2 features it would be nice to have an option to completely disable them.

And also, would it be possible to add settings to change the random variables of the bot? I am aware changing these could get the bot recognised by the admins, but I would just like to be able to see the max/min possible times between farms so that I can optimise the "bot update time" to be the shortest possible without farms not being ready. Even better would be if the bot could choose this time itself to make sure farms are always ready, but this might be a lot of work, so I understand if that is not possible (sorry I don't know much about coding lol)

Thanks again for your work on this and sharing it! :)

**Comment by BrandonWill** [comment]
@NS58 

Yes, all of these are possible. The bot didn't always have randomization and it was never detected then. It's just good practice to add it. In my experience through some very wild testing (seeing what I can get away with), there is very little bot detection. In the past, I made some modifications and farmed 40 cities under 1 second for several months without being banned. The only way I've learned you can be banned is from:

1) Making a call that no human ever would. For instance, when you first login to your world you obtain a lot of information. Repeating that call while you're logged in will tag you. This ban will hit you within a few days.

2) Being reported for botting by someone else in the game.

Nonetheless, these suggestions will have to wait for some time. I'll be posting the next update in the next hour or two for automatic academy, which is sure to slow you down a bit as well :).

**Comment by BrandonWill** [comment]
The newest version of the bot is now out: 0.24. Please read the below before updating.

Upon updating, please double check all of your save files. There was a bug in the previous versions where it would write more than it should've to your save file the first time it was saved. Due to the way loading/saving worked then, it didn't matter and nobody really noticed including me. When I was adding the academy, I noticed this bug and corrected it. This only affected the barracks/docks save files so your troops may not be correct. This was also a cause of a small memory leak, so you should see some increased performance upon upgrading.

I've spent about a month testing and correcting issues with the automatic academy portion, so please let me know if you see any issues with it. The Building Panel has also been updated in this version. Some other tweaks have been made and are in the change log.

**Comment by Frank18000** [comment]
this update is working great. Good job.

Are you still looking at the attack option?

**Comment by BrandonWill** [comment]
Attack is still being looked into for a pretty implementation in my free time

**Comment by lusuma** [comment]
Hello
With farming:
When the warehouse is full, do not continue to collect resources from the villages. If there is more than one city on the island, when one of them has a full store, go collect resources from the villages, the other city.

**Comment by tepacz** [comment]
Amazing job with the bot, thank you so much for it. You're a legend mate.

**Comment by zaltan13** [comment]
Hello, first of all thanks for your effort, you are doing a GREAT job. I'm having a problem when trying to open version 24: "A JNI error has occurred, please check your installation and try again". I have no problem with V22, any idea?

**Comment by tepacz** [comment]
Hi, the bot keeps crashing. Any idea on how to fix that? 
Also I can't run the new version of it

**Comment by BrandonWill** [comment]
The newest version of the bot uses Java 9. You can find the download for it here: http://www.oracle.com/technetwork/java/javase/downloads/java-archive-javase9-3934878.html 

 It should work with Java SE Runtime Environment 9.0.4.

Sadly, Oracle completely gave up on Java 9 and has moved onto Java 10. Most Java versions are usually updated for around 5 years, and this version only lasted a year. They broke tons of things with Java 9 & that required a lot of changes for the bot. I will have to plan for the java 10 upgrade

**Comment by tepacz** [comment]
Cheers bud, it works perfectly now. Thanks a lot for the bot.

**Comment by azaracla** [comment]
Hi Brandon,

First of all, great work, I discovered your bot a few days ago because I am working myself on a Python bot (I know Java, but I a Python fan ! )

I have one issue you might be able to help me with : some requests to the Grepolis servers require in the string query a parameter called "_" that is unique and increment every request (similar to the timestamp,  but longer, e.g 1526989364125) 
I can't manage to get response data to requests such as map_data, or the request to farm towns, etc without this parameter... 
I don't see this parameter in your requests, how do you manage to get theses requests through ? 

TY in advance, 
A.

**Comment by abelillanes** [comment]
Hi Brandon.

Where can I look for the compilation manual? I don't see any main.java file or something like this to compile. Also, which compiler did you use?

Thanks in advance.
Regards.

**Comment by BrandonWill** [comment]
@azaracla

This data is read on startup when it first logs in. This bot does not force read any data from Grepolis through any unhuman-like requests. 

The parameter after  "_" is just the time EPOCH time in UTC. I have never needed this to make calls, but I do use it in some scenarios such as to calculate if premium farmers are going to expire, etc.

    public static long getServerUnixTime() {
        TimeZone tz = TimeZone.getTimeZone("UTC");
        Calendar c = Calendar.getInstance(tz);
        return c.getTimeInMillis() / 1000;
    }

@Abelillanes,

The "main.java" file is the GrepolisBot.java. I use intellij to develop. The bot is currently using Java 9 JDK. 

Java 10 was released already and the bot wasn't compatible which took over a month of developing and then reverting.

However, this will soon be deprecated. As of Java 11, people won't be able to download the Java JRE anymore. Oracle is expecting developers to create a JRE and distribute it with the program. They are also removing JavaFX with Java 11, which is what I use for the browser.

With that said, this has greatly delayed any actual progress on the bot.  I have successfully ported the bot to Java 11 and added the JavaFX libraries. I still need to learn how to utilize jlink to create the JRE to distribute. 

Once I ensure the bot is still going to be usable, I'll continue on the old projects such as auto completing the tutorial (which will also be testing the auto attack feature).

**Comment by ThomasCloarec** [comment]
Hello @BrandonWill 

Thanks for your work, the bot works well.

I just found three problems while running the last version of it (all about research images) : 
- All the research images files should be in lowercase as your program asks for images with names in lowercase (so actually, with the last version of the bot, no images are shown on the research panel).
- The image for the research (stone_hail) is missing in files.
- In the program, you're sometimes calling it "stone_storm" while it should only be "stone_hail" (not a big problem though but stills need to be fixed)

Are you still working on it ? If yes, I would be happy to help you on it, I got some knowledge in java but I  would like to talk a little bit with you to understand on what are your actually working to not work for nothing (maybe by PM ?).

Thanks,
Thomas

**Comment by BrandonWill** [comment]
@ThomasCloarec 

1. I think this is something github has done? It hasn't always been that way at least. Locally, these haven't been updated since 2016 as well. 
![image](https://user-images.githubusercontent.com/949047/48979437-c3714900-f088-11e8-8796-cce26eb46e76.png)

2. Same as above. Not too sure where it went. I'll make sure it's added in next release. The location will change for a lot of files.

3. You'll see a lot of things named differently internally in the bot. This is because Grepolis calls them a different name in the backend. I've stuck to their naming conventions in the backend to make development easier.

I try to work on it when I can. The current project is moving it off of Java 9 (which has ended support and downloads will disappear soon) on over to Java 11. The bot is also utilizing Maven framework now to make that port easier. I'm always open to help on the bot. Everything in Grepolis can be automated if you spend the time on it. GitHub remove private messaging some time back. You can reach out to my email for the bot as a starting place: githubgrepolisbot@gmail.com

**Comment by ThomasCloarec** [comment]
@BrandonWill 
Okay thanks, 
to be able to help you, I will need to understand all your code (actually the part about getting all the information from Grepolis is a bit complicated for me), also I still need to learn how does Marven works.

I will contact you soon, will probably wait for the next release as a lot of things will change.

Thanks,
Thomas

**Comment by matz98** [comment]
Hey, what do you think about an auto Trader with farmers?

Exaple: you have 50 woods, 1000 stone, 100 silver.
Now the bot could trade all the values at the same amount.
50+1000+100/3 = 383
means:
383-50= 333 means: give 333 stone for woods
100-50 = 50 means: give 50 stone for silver

**Comment by matz98** [comment]
You also could use the factor, for example:
only trade if  trade factor is >than x   or you use the factor in the mathematical formula:

x=1,25

383-50 = 333/x(trade factor 1.25), so you alsways have the correct amount. means: give 266 stone and you get 333 woods
